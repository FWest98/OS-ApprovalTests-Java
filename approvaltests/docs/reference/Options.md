<a id="top"></a>

# Options

<!-- toc -->
## Contents

* [Introduction](#introduction)
* [Fluent Interface](#fluent-interface)
* [Reporters](#reporters)
* [Scrubbers](#scrubbers)
* [File Options](#file-options)
    * [File Extensions](#file-extensions)
* [Defaults](#defaults)
* [Adding to Existing Options object](#adding-to-existing-options-object)<!-- endToc -->

## Introduction

There are many things you might want to tweak with Approval Tests. `Options` is the entry-point for many of the changes.
It is on all `verify()` methods, as an optional parameter.

## Fluent Interface

`Options` utilizes a fluent interface, allowing you to chain together commands. Each returned object is a new copy.

<!-- snippet: specify_all_the_options -->
<a id='snippet-specify_all_the_options'></a>
```cpp
ApprovalTests::Options()
    .withReporter(ApprovalTests::QuietReporter())
    .withScrubber(ApprovalTests::Scrubbers::scrubGuid)
    .fileOptions().withFileExtension(".json")
```
<sup><a href='/tests/DocTest_Tests/DocTestApprovalTestTests.cpp#L50-L55' title='Snippet source file'>snippet source</a> | <a href='#snippet-specify_all_the_options' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## Reporters

[Reporters](/doc/Reporters.md#top) launch diff tools upon failure.  Read how to configure them with `Options` [here](../Reporters.md#via-options):

## Scrubbers

[Scrubbers](/doc/explanations/Scrubbers.md#top) clean output to help remove inconsistent pieces of text, such as dates.  Read how to set up Scrubbers [here](../Scrubbers.md#configuring-scrubbers):
There are two ways to set a Scrubber.

## File Options

The `Options::FileOptions` class exists to customise the `.approved` and `.received` files in various ways.

For now, it just controls the file extension.

### File Extensions

If you want to change the file extension of both the approved and received files, use `withFileExtension()`.

<!-- snippet: basic_approval_with_file_extension -->
<a id='snippet-basic_approval_with_file_extension'></a>
```cpp
using namespace ApprovalTests;

Approvals::verify("text to be verified",
                  Options().fileOptions().withFileExtension(".xyz"));
```
<sup><a href='/tests/DocTest_Tests/core/OptionsExamples.cpp#L7-L12' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic_approval_with_file_extension' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

**Note:** `withFileExtension()` returns an `Options` object, so it's possible to keep appending more `with...()` calls.

## Defaults

The default constructor for Options does:

* no scrubbing
* uses file extension `.txt`
* uses whatever is currently set as [the default reporter](#registering-a-default-reporter).

## Adding to Existing Options object

Because of the fluent options, you can modify a specific option from an existing Options object,
while retaining all other settings, and not changing the original object.

```cpp
verifyWithQuietReporter(std::string text, const Options& o)
{
    Approvals::verify(text, o.withReporter(QuietReporter());
}
```

---

[Back to User Guide](/doc/README.md#top)