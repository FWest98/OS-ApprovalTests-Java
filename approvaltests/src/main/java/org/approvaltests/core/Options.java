package org.approvaltests.core;

import com.spun.util.ArrayUtils;
import org.approvaltests.Approvals;
import org.approvaltests.ReporterFactory;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.NamerWrapper;
import org.approvaltests.scrubbers.NoOpScrubber;
import org.approvaltests.writers.ApprovalWriterFactory;
import org.approvaltests.writers.DefaultApprovalWriterFactory;

import java.util.HashMap;
import java.util.Map;

public class Options
{
  private enum Fields {
                       SCRUBBER, REPORTER, FILE_OPTIONS_FILE_EXTENSION, FILE_OPTIONS_NAMER, WRITER;
  }
  private final Map<Fields, Object> fields = new HashMap<>();
  public Options()
  {
  }
  public Options(Scrubber scrubber)
  {
    fields.put(Fields.SCRUBBER, scrubber);
  }
  public Options(ApprovalFailureReporter reporter)
  {
    fields.put(Fields.REPORTER, reporter);
  }
  private Options(Map<Fields, Object> fields, Fields key, Object value)
  {
    this.fields.putAll(fields);
    this.fields.put(key, value);
  }
  public ApprovalFailureReporter getReporter()
  {
    return ArrayUtils.getOrElse(fields, Fields.REPORTER, ReporterFactory::get);
  }
  public Options withReporter(ApprovalFailureReporter reporter)
  {
    return new Options(fields, Fields.REPORTER, reporter);
  }
  public Options withScrubber(Scrubber scrubber)
  {
    return new Options(fields, Fields.SCRUBBER, scrubber);
  }
  public String scrub(String input)
  {
    return getScrubber().scrub(input);
  }
  private Scrubber getScrubber()
  {
    return ArrayUtils.getOrElse(fields, Fields.SCRUBBER, () -> NoOpScrubber.INSTANCE);
  }
  public ApprovalWriter createWriter(Object o)
  {
    ApprovalWriterFactory factory = ArrayUtils.getOrElse(fields, Fields.WRITER,
        DefaultApprovalWriterFactory::getDefaultFactory);
    return factory.create(o, this);
  }
  public Options withWriter(ApprovalWriterFactory approvalWriterFactory)
  {
    return new Options(fields, Fields.WRITER, approvalWriterFactory);
  }
  public FileOptions forFile()
  {
    return new FileOptions(this.fields);
  }
  public static class FileOptions
  {
    private final Map<Fields, Object> fields;
    public FileOptions(Map<Fields, Object> fields)
    {
      this.fields = fields;
    }
    public Options withExtension(String fileExtensionWithDot)
    {
      if (!fileExtensionWithDot.startsWith("."))
      {
        fileExtensionWithDot = "." + fileExtensionWithDot;
      }
      return new Options(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, fileExtensionWithDot);
    }
    public ApprovalNamer getNamer()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_NAMER, Approvals::createApprovalNamer);
    }
    public String getFileExtension()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, () -> ".txt");
    }
    public Options withBaseName(String fileBaseName)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      return new Options(fields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
    public Options withName(String fileBaseName, String extension)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      HashMap<Fields, Object> newFields = new HashMap<>(fields);
      newFields.put(Fields.FILE_OPTIONS_FILE_EXTENSION, extension);
      return new Options(newFields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
  }
}
