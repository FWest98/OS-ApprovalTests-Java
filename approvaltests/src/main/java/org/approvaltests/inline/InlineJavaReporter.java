package org.approvaltests.inline;

import com.spun.util.StringUtils;
import com.spun.util.io.FileUtils;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalReporterWithCleanUp;
import org.approvaltests.namer.StackTraceNamer;

import java.io.File;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class InlineJavaReporter implements ApprovalFailureReporter, ApprovalReporterWithCleanUp
{
  private final String                  sourceFilePath;
  private final StackTraceNamer         stackTraceNamer;
  private final ApprovalFailureReporter reporter;
  public InlineJavaReporter(ApprovalFailureReporter reporter)
  {
    this.reporter = reporter;
    this.stackTraceNamer = new StackTraceNamer();
    this.sourceFilePath = stackTraceNamer.getSourceFilePath();
  }
  public String getSourceFilePath()
  {
    return sourceFilePath;
  }
  @Override
  public boolean report(String received, String approved)
  {
    String sourceFile = sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".java";
    String newSource = createReceived(FileUtils.readFile(received));
    return reporter.report(newSource, sourceFile);
  }
  public String createReceived(String actual)
  {
    String file = sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".java";
    String received = getReceivedFileName();
    String text = FileUtils.readFile(file);
    String fullText = createNewReceivedFileText(text, actual, this.stackTraceNamer.getInfo().getMethodName());
    FileUtils.writeFile(new File(received), fullText);
    return received;
  }
  private String getReceivedFileName()
  {
    return sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".received.txt";
  }
  public static String createNewReceivedFileText(String text, String actual, String methodName)
  {
    text = text.replaceAll("\r\n", "\n");
    CodeParts codeParts = CodeParts.splitCode(text, methodName);
    if (codeParts.method.contains("expected = \"\"\""))
    {
      replaceExpected(codeParts, actual);
    }
    else
    {
      addExpected(codeParts, actual);
    }
    return codeParts.getFullCode();
  }
  private static void addExpected(CodeParts codeParts, String actual)
  {
    int start = codeParts.method.indexOf("{") + 2;
    String before = codeParts.method.substring(0, start);
    String after = codeParts.method.substring(start);
    codeParts.method = before + getExpected(actual, codeParts.tab) + after;
  }
  private static String getExpected(String actual, String tab)
  {
    return String.format("%s%svar expected = \"\"\"\n%s%s%s%s\"\"\";\n", tab, tab, indent(actual, tab), tab, tab,
        tab);
  }
  private static void replaceExpected(CodeParts codeParts, String actual)
  {
    int start = codeParts.method.indexOf("expected = \"\"\"");
    start = codeParts.method.substring(0, start).lastIndexOf("\n") + 1;
    int end = codeParts.method.indexOf("\"\"\";");
    end = codeParts.method.indexOf("\n", end) + 1;
    String before = codeParts.method.substring(0, start);
    String after = codeParts.method.substring(end);
    codeParts.method = before + getExpected(actual, codeParts.tab) + after;
  }
  public static String indent(String actual, String tab)
  {
    String[] split = StringUtils.split(actual, "\n");
    String output = "";
    for (String line : split)
    {
      output += tab + tab + tab + line + "\n";
    }
    return output;
  }
  private static String extractLeadingWhitespace(String text)
  {
    Pattern pattern = Pattern.compile("^\\s+");
    Matcher matcher = pattern.matcher(text);
    if (matcher.find())
    { return matcher.group(); }
    return "\t";
  }
  @Override
  public void cleanUp(String received, String approved)
  {
    FileUtils.delete(getReceivedFileName());
  }
}
