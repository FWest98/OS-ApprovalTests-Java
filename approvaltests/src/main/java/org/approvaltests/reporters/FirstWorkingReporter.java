package org.approvaltests.reporters;

import org.approvaltests.core.ApprovalFailureReporter;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.util.List;

public class FirstWorkingReporter implements EnvironmentAwareReporter
{
  private final EnvironmentAwareReporter[] reporters;
  public FirstWorkingReporter(EnvironmentAwareReporter... reporters)
  {
    this.reporters = reporters;
  }
  public static FirstWorkingReporter combine(EnvironmentAwareReporter front, ApprovalFailureReporter last)
  {
    return new FirstWorkingReporter(front, wrap(last));
  }
  private static EnvironmentAwareReporter wrap(ApprovalFailureReporter last)
  {
    if (last instanceof EnvironmentAwareReporter)
    { return (EnvironmentAwareReporter) last; }
    return new AlwaysWorkingReporter(last);
  }
  @Override
  public void report(String received, String approved)
  {
    for (EnvironmentAwareReporter reporter : reporters)
    {
      if (reporter.isWorkingInThisEnvironment(received))
      {
        reporter.report(received, approved);
        return;
      }
    }
  }
  @Override
  public boolean isWorkingInThisEnvironment(String forFile)
  {
    for (EnvironmentAwareReporter reporter : reporters)
    {
      if (reporter.isWorkingInThisEnvironment(forFile))
      { return true; }
    }
    return false;
  }
  public EnvironmentAwareReporter[] getReporters()
  {
    return reporters;
  }
  /**
   *
   * @deprecated Use {@link #getWorkingReportersForEnvironment()} instead.}
   */
  @Deprecated
  public List<EnvironmentAwareReporter> getWorkingReportersForEnviroment()
  {
    return getWorkingReportersForEnvironment();
  }
  public List<EnvironmentAwareReporter> getWorkingReportersForEnvironment()
  {
    return Query.where(reporters, new Function1<EnvironmentAwareReporter, Boolean>()
    {
      public Boolean call(EnvironmentAwareReporter r)
      {
        return r.isWorkingInThisEnvironment("a.txt");
      }
    });
  }
  @Override
  public String toString()
  {
    return getClass().getName();
  }
}
