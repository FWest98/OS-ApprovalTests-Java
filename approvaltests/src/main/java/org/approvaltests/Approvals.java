package org.approvaltests;

import com.spun.util.*;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.*;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class Approvals
{
  public static Loader<ApprovalNamer> namerCreater = new Loader<ApprovalNamer>()
  {
    public ApprovalNamer load()
    {
      return new StackTraceNamer();
    }
  };
  public static void verify(String response)
  {
    verify(new ApprovalTextWriter(response, "txt"));
  }
  public static void verify(Object o)
  {
    verify("" + o);
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }

  public static void verifyHtml(String response)
  {
    verify(new ApprovalTextWriter(response, "html"));
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile));
  }

  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), reporter);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, createApprovalNamer(), getReporter());
  }
  public static void verifyXml(String xml)
  {
    verify(new ApprovalXmlWriter(xml));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, getReporter());
  }
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    if (!approver.approve())
    {
      approver.reportFailure(reporter);
      approver.fail();
    }
    else
    {
      approver.cleanUpAfterSuccess(reporter);
    }
  }
  public static void verify(ExecutableQuery query)
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), createApprovalNamer(),
        new ExecutableQueryFailure(query));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"));
  }
  public static void verify(ResultSet rs)
  {
    verify(new ResultSetApprovalWriter(rs));
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader));
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }

  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + File.separator + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory));
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  public static void verifyJson(String json)
  {
    verify(JsonUtils.prettyPrint(json), "json");
  }
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(new ApprovalTextWriter("" + actual, fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  public static void verifyAsJson(Object o)
  {
    verify(JsonUtils.asJson(o), "json");
  }
  public static void verifyException(Action0 runnableBlock)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null) { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()));
  }
}