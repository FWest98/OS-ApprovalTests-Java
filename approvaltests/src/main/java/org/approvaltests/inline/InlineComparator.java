package org.approvaltests.inline;

import com.spun.util.io.FileUtils;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.core.Options;
import org.approvaltests.core.VerifyResult;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.DiffMergeReporter;
import org.approvaltests.writers.ApprovalWriterFactory;
import org.lambda.functions.Function2;

import java.io.File;
import java.io.IOException;

import static org.approvaltests.writers.Writer.approved;
import static org.approvaltests.writers.Writer.received;

public class InlineComparator
    implements
      Function2<File, File, VerifyResult>,
      ApprovalNamer,
      ApprovalWriter,
      ApprovalWriterFactory,
      ApprovalFailureReporter
{
  private final String          sourceFilePath;
  private final StackTraceNamer stackTraceNamer;
  private String                expected;
  private String                actual;
  private File                  approvedFile;
  private File                  receivedFile;
  public int fileWrites = 0;
  public InlineComparator(String expected)
  {
    this.expected = expected;
    stackTraceNamer = new StackTraceNamer();
    sourceFilePath = stackTraceNamer.getSourceFilePath();
  }
  @Override
  public VerifyResult call(File received, File approved)
  {
    if (expected.equals(actual))
    {
      return VerifyResult.SUCCESS;
    }
    else
    {
      writeFiles();
      return VerifyResult.FAILURE;
    }
  }

  private void writeFiles() {
    FileUtils.writeFile(approvedFile, expected);
    fileWrites++;
    FileUtils.writeFile(receivedFile, actual);
    fileWrites++;
  }

  @Override
  public File getApprovedFile(String extensionWithDot)
  {
    if (approvedFile == null)
    {
      try
      {
        this.approvedFile = File.createTempFile("temp", approved+extensionWithDot);
      }
      catch (IOException e)
      {
        throw new RuntimeException(e);
      }
    }
    return approvedFile;
  }
  @Override
  public File getReceivedFile(String extensionWithDot)
  {
    if (receivedFile == null)
    {
      try
      {
        this.receivedFile = File.createTempFile("temp", received+extensionWithDot);
      }
      catch (IOException e)
      {
        throw new RuntimeException(e);
      }
    }
    return receivedFile;
  }
  @Override
  public ApprovalNamer addAdditionalInformation(String info)
  {
    return this;
  }
  @Override
  public String getApprovalName()
  {
    return "";
  }
  @Override
  public String getSourceFilePath()
  {
    return sourceFilePath;
  }
  @Override
  public File writeReceivedFile(File received)
  {
    return received;
  }
  @Override
  public String getFileExtensionWithDot()
  {
    return ".txt";
  }
  @Override
  public ApprovalWriter create(Object content, Options options)
  {
    this.actual = options.scrub("" + content);
    if (!actual.endsWith("\n"))
    {
      this.actual += "\n";
    }
    return this;
  }
  @Override
  public boolean report(String received, String approved)
  {
    ApprovalFailureReporter reporter = new Options().getReporter();
    String sourceFile = sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".java";
    String newSource = createReceived();
    return reporter.report(newSource, sourceFile);
  }

  private String createReceived()
  {
    String file = sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".java";
    String received = sourceFilePath + stackTraceNamer.getInfo().getClassName() + ".received.txt";
    String text = FileUtils.readFile(file);
    String fullText = createNewReceivedFileText(text, this.actual, this.stackTraceNamer.getInfo().getMethodName());
    FileUtils.writeFile(new File(received), fullText);
    return received;
  }

  public static String createNewReceivedFileText(String text, String actual, String methodName) {
    int start = text.indexOf("void " + methodName + "(");
    start = text.indexOf("{", start);
    int next = text.indexOf("\n", start);
    int end = text.indexOf("}", next);
    int endString = text.indexOf("\"\"\";", next);
    String part1 = text.substring(0, next);
    String part2 = null;
    if (0 < endString && endString < end)
    {
      part2 = text.substring(endString + 4);
    }
    else
    {
      part2 = text.substring(next);
    }
    String fullText = String.format("%s\n\t\tvar expected = \"\"\"\n%s\t\t\"\"\";\n%s", part1, indent(actual),
        part2);
    return fullText;
  }

  public static String indent(String actual)
  {
    String[] split = actual.split("\n");
    String output = "";
    for (String line : split)
    {
      output += "\t\t" + line + "\n";
    }
    return output;
  }
}
