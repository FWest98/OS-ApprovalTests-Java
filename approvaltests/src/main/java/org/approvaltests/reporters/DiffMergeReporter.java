package org.approvaltests.reporters;

import org.approvaltests.reporters.linux.DiffMergeLinuxReporter;
import org.approvaltests.reporters.macosx.DiffMergeMacOsReporter;

public class DiffMergeReporter extends FirstWorkingReporter
{
  public static final DiffMergeReporter INSTANCE = new DiffMergeReporter();
  public DiffMergeReporter()
  {
    super(DiffMergeMacOsReporter.INSTANCE, DiffMergeLinuxReporter.INSTANCE);
  }
}
