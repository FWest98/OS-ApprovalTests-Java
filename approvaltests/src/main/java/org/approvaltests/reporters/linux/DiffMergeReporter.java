package org.approvaltests.reporters.linux;

import org.approvaltests.reporters.DiffPrograms.Linux;
import org.approvaltests.reporters.GenericDiffReporter;

public class DiffMergeReporter extends GenericDiffReporter
{
  public static final DiffMergeReporter INSTANCE = new DiffMergeReporter();
  public DiffMergeReporter()
  {
    super(Linux.DIFF_MERGE);
  }
}