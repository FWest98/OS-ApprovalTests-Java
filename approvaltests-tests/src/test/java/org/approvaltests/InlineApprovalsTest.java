package org.approvaltests;

import org.approvaltests.core.Options;
import org.approvaltests.inline.InlineComparator;
import org.approvaltests.inline.InlineJavaReporter;
import org.approvaltests.reporters.*;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledOnJre;
import org.junit.jupiter.api.condition.JRE;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

@EnabledOnJre({JRE.JAVA_21})
public class InlineApprovalsTest
{
  @Test
  @UseReporter(DiffMergeReporter.class)
  public void testWithBuiltinReporter()
  {
    var expected = """
        Hello There***
        """;
    Options inline =
        // begin-snippet: inline_approvals
        new Options().inline(expected);
    // end-snippet
    Approvals.verify("Hello There***", inline);
  }
  @UseReporter(DiffMergeReporter.class)
  @Test
  public void testWithSpecificReporter()
  {
    var expected = """
        Hello Lada***
        """;
    Options inline = new Options().inline(expected).withReporter(DiffMergeReporter.INSTANCE);
    Approvals.verify("Hello Lada***", inline);
  }
  @Test
  public void testCreateReceivedFileText()
  {
    var inputs = List.of("""
              @Test
              public void testyMctest() {
                var expected = ""\"
                    Hello World***
                    ""\";
                Approvals.verify("", Options.inline(expected));
              }
        """, """
             @Test
             public void testyMctest()
             {
               Approvals.verify("", Options.inline(expected));
             }
        """, """
             @Test
             public void testyMctest()
             {
               var expected = ""\"
                   Hello World***
                   ""\";
               Approvals.verify("", Options.inline(expected));
             }
        """, """
              @Test
              public void testyMctest(int foo) {
                var expected = ""\"
                    Hello World***
                    ""\";
                Approvals.verify("", Options.inline(expected));
              }
        """);
    Approvals.verifyAll("Substitution", inputs, i -> "******\n" + i + "\nBecomes:\n"
                                                     + InlineJavaReporter.createNewReceivedFileText(i, "1\n2", "testyMctest"));
  }
  @Test
  @UseReporter(QuietReporter.class)
  public void testReportingCode()
  {
    Options inlineWithCode = new Options().inline("", true);
    Options inlineNoCode = new Options().inline("", false);
    var resultWithCode = inlineWithCode.getReporter();
    assertEquals(InlineComparator.class, resultWithCode.getClass());
    assertEquals(QuietReporter.class,
        ((FirstWorkingReporter) inlineNoCode.getReporter()).getReporters()[1].getClass());
  }
}
