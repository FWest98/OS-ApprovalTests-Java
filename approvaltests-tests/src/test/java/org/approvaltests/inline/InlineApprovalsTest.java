package org.approvaltests.inline;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.DiffMergeReporter;
import org.approvaltests.reporters.FirstWorkingReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class InlineApprovalsTest
{
  @Test
  @UseReporter(DiffMergeReporter.class)
  public void testWithBuiltinReporter()
  {
    var expected = """
        Hello There***
        """;
    Options inline =
        // begin-snippet: inline_approvals
        new Options().inline(expected);
    // end-snippet
    Approvals.verify("Hello There***", inline);
  }
  @UseReporter(DiffMergeReporter.class)
  @Test
  public void testWithSpecificReporter()
  {
    var expected = """
        Hello Lada***
        """;
    Options inline = new Options().inline(expected).withReporter(DiffMergeReporter.INSTANCE);
    Approvals.verify("Hello Lada***", inline);
  }
  @Test
  public void testCreateReceivedFileText()
  {
    var inputs = List.of("""
          @Test
          public void testyMctest() {
            var expected = ""\"
                Hello World***
                ""\";
            Approvals.verify("", Options.inline(expected));
          }
        """, """
        \t@Test
        \tpublic void testyMctest()
        \t{
        \t\tApprovals.verify("", Options.inline(expected));
        \t}
        """, """
          @Test
          public void testyMctest()
          {
            var expected = ""\"
                Hello World***
                ""\";
            Approvals.verify("", Options.inline(expected));
          }
        """, """
          @Test
          public void testyMctest(int foo) {
            var expected = ""\"
                Hello World***
                ""\";
            Approvals.verify("", Options.inline(expected));
          }
        """, """
          @Test
          void testyMctest() {
              Customer customer = Customer.create()
                  .existingCustomer()
                  .speakingEnglish()
                  .build();
              stateService.save(TestUtils.userId().build(), customer);
              var expected = ""\"
                  [Customer]: account lookup
                  [     Bot]: Hi there!
                  [     Bot]: Let me try to help.
                  [     Bot]: routes to '12345'
                  ""\";
              verifyConversation(expected, "account lookup");
          }
        """);
    Approvals.verifyAll("Substitution", inputs, i -> "******\n" + i + "\nBecomes:\n"
        + InlineJavaReporter.createNewReceivedFileText(i, "1\n2", "testyMctest"));
  }
  @Test
  @UseReporter(QuietReporter.class)
  public void testReportingCode()
  {
    Options inlineWithCode = new Options().inline("", true);
    Options inlineNoCode = new Options().inline("", false);
    var resultWithCode = inlineWithCode.getReporter();
    assertEquals(InlineJavaReporter.class, resultWithCode.getClass());
    assertEquals(QuietReporter.class,
        ((FirstWorkingReporter) inlineNoCode.getReporter()).getReporters()[1].getClass());
  }
  @Test
  void testEmptyLineAtTheEnd()
  {
    var expected = """
        Jeff Jeffty Jeff
        born on Jeffteen of Jeff, Nineteen-eighty-Jeff

        """;
    Approvals.verify(greet("Jeff"), new Options().inline(expected));
  }
  private String greet(String name)
  {
    return """
        %s %sty %s
        born on %steen of %s, Nineteen-eighty-%s

        """.formatted(name, name, name, name, name, name);
  }
}
