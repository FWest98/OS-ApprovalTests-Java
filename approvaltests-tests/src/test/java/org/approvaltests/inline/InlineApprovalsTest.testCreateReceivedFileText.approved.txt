Substitution


******
  @Test
  public void testyMctest() {
    var expected = """
        Hello World***
        """;
    Approvals.verify("", Options.inline(expected));
  }

Becomes:
  @Test
  public void testyMctest() {
    var expected = """
      1
      2
      """;
    Approvals.verify("", Options.inline(expected));
  }

******
  @Test
  public void testyMctest () {
    var expected = """
        Hello World***
        """;
    Approvals.verify("", Options.inline(expected));
  }

Becomes:
  @Test
  public void testyMctest () {
    var expected = """
      1
      2
      """;
    Approvals.verify("", Options.inline(expected));
  }

******
	@Test
	public void testyMctest()
	{
		Approvals.verify("", Options.inline(expected));
	}

Becomes:
	@Test
	public void testyMctest()
	{
		var expected = """
			1
			2
			""";
		Approvals.verify("", Options.inline(expected));
	}

******
  @Test
  public void testyMctest()
  {
    var expected = """
        Hello World***
        """;
    Approvals.verify("", Options.inline(expected));
  }

Becomes:
  @Test
  public void testyMctest()
  {
    var expected = """
      1
      2
      """;
    Approvals.verify("", Options.inline(expected));
  }

******
  @Test
  public void testyMctest(int foo) {
    var expected = """
        Hello World***
        """;
    Approvals.verify("", Options.inline(expected));
  }

Becomes:
  @Test
  public void testyMctest(int foo) {
    var expected = """
      1
      2
      """;
    Approvals.verify("", Options.inline(expected));
  }

******
  @Test
  void testyMctest() {
      Customer customer = Customer.create()
          .existingCustomer()
          .speakingEnglish()
          .build();
      stateService.save(TestUtils.userId().build(), customer);
      var expected = """
          [Customer]: account lookup
          [     Bot]: Hi there!
          [     Bot]: Let me try to help.
          [     Bot]: routes to '12345'
          """;
      verifyConversation(expected, "account lookup");
  }

Becomes:
  @Test
  void testyMctest() {
      Customer customer = Customer.create()
          .existingCustomer()
          .speakingEnglish()
          .build();
      stateService.save(TestUtils.userId().build(), customer);
    var expected = """
      1
      2
      """;
      verifyConversation(expected, "account lookup");
  }

